python/ray/ml/train/data_parallel_trainer.py

```diff
@@ -292,27 +292,36 @@ def training_loop(self) -> None:
         else:
             resume_checkpoint_dict = None
 
-        # Tell Ray Train to only shard the train dataset and not the other datasets.
-        # This is purely an implementation detail and users do not need to know about
-        # this.
-        # TODO(amog): Refactor this to remove hack and make this more modular.
-        #  TrainingIterator should accept a generic custom_ingest_func that contains
-        #  the logic for how to split the Datasets.
-        updated_dataset_dict = {}
-        for key, value in self.datasets.items():
-            if key == TRAIN_DATASET_KEY:
-                updated_dataset_dict[key] = value
-            else:
-                # Ray Train will strip out the added string before exposing to users.
-                updated_dataset_dict[key + "_NO-SHARD"] = value
+        def dataset_split_fn(dataset_dict, training_worker_handles):
+            dataset_dict_splits = [{} for _ in range(len(training_worker_handles))]
+
+            for key, dataset in dataset_dict.items():
+                if key == TRAIN_DATASET_KEY:
+                    dataset_splits = dataset.split(
+                        len(training_worker_handles),
+                        equal=True,
+                        locality_hints=training_worker_handles,
+                    )
+                else:
+                    # Only shard the training dataset.
```

Comment:
Could you add a comment explaining why we're only sharding the training dataset?

Reply:
Added!